h5py/_hl/dataset.py

```diff
@@ -242,6 +242,86 @@ def __exit__(self, *args):
             # pylint: disable=protected-access
             self._dset._dxpl.set_dxpl_mpio(h5fd.MPIO_INDEPENDENT)
 
+class ChunkIterator(object):
+    """
+    Class to iterate through list of chunks of a given dataset
+    # Note: Only works with PY3 since as is doesn't have a "next = __next__" redef
+    """
+    def __init__(self, dset, source_sel=None):
+        self._shape = dset.shape
+        rank = len(dset.shape)
+
+        if not dset.chunks:
+            # can only use with chunked datasets
+            raise TypeError("Chunked dataset required")
+
+        self._layout = dset.chunks
+        if source_sel is None:
+            # select over entire dataset
+            slices = []
+            for dim in range(rank):
+                slices.append(slice(0, self._shape[dim]))
+            self._sel = tuple(slices)
+        else:
+            if isinstance(source_sel, slice):
+                self._sel = (source_sel,)
+            else:
+                self._sel = source_sel
+        if len(self._sel) != rank:
+            raise ValueError("Invalid selection")
+        self._chunk_index = []
+        for dim in range(rank):
+            s = self._sel[dim]
+            if s.start < 0 or s.stop > self._shape[dim] or s.stop <= s.start:
+                raise ValueError("Invalid selection")
+            index = s.start // self._layout[dim]
+            self._chunk_index.append(index)
+
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+
+        def get_ret(item):
+            if len(item) == 1:
+                return item[0]
+            else:
+                return tuple(item)
```

Comment:
I'd say let's return a tuple in all cases - a 1-tuple is equally valid for slicing, and it's easy to unpack if someone knows it's a 1D dataset and wants a slice:

```python
for sl, in ds.iter_chunks():
    pass
```

Reply:
I've updated to always return tuples.
Also added rst doc.